lib/Sema/CSApply.cpp

```diff
@@ -304,9 +252,71 @@ Solution::resolveConcreteDeclRef(ValueDecl *decl,
         thunk->setBodySynthesizer(synthesizeForwardingThunkBody, cast<FuncDecl>(newDecl));
         thunk->setSelfAccessKind(fn->getSelfAccessKind());
 
-        return ConcreteDeclRef(thunk);
+        newDecl = thunk;
+      }
+    }
+
+    if (auto fn = dyn_cast<AbstractFunctionDecl>(newDecl)) {
```

Comment:
@hyp do you think it would be good to factor this if/hack out into it's own function?